{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pandas.api.types import is_string_dtype\n",
    "from pandas.api.types import is_numeric_dtype\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in the data frame, drop the unnecessary index row & transpose it so rows are loci and columns are individuals.\n",
    "df = pd.read_parquet(\"~/Documents/UMontana/Research/YNP/AHQsd/AHQsd_analyses/AHQsd_genotypes_filt.parquet\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# label by what linkage group the loci belong to\n",
    "tmpDF = pd.DataFrame(columns=['chr','site'])\n",
    "tmpDF[['chr','site']] = df['index'].str.split('_', expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['chr'] = tmpDF['chr']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['site'] = tmpDF['site']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# split them\n",
    "for chromo in range(1,15):\n",
    "    lg_chromo = df[df.chr == str(chromo)]\n",
    "    lg_chromo.to_csv(f'data/lg_chromo_{chromo}.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "tdf = df.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def list_name_indivs(ds):\n",
    "    return ds.columns.to_list()[2:323] \n",
    "\n",
    "def create_bins(subset_ds):\n",
    "    # create bins of 18 SNPs each in each linkage group\n",
    "    bins = [i for i in np.arange(subset_ds.index.min(), subset_ds.index.max(), 18)] + [np.inf]\n",
    "    return bins\n",
    "\n",
    "def create_stats_template(ds, chromo_number):\n",
    "    # create dataframes for the loop to write to. \n",
    "    return pd.DataFrame(columns=[\"indv_name\",\"count\",\"mean\",f\"num_of_nans_chr{chromo_number}\"])\n",
    "\n",
    "def cut_by_bins(subset_ds, bins):\n",
    "    return pd.cut(subset_ds.index,bins=bins)\n",
    "\n",
    "def create_stats_dataframe(small_list, ds, bins, chromo_number):\n",
    "    stats_dataframe = create_stats_template(ds, chromo_number)\n",
    "    chromo_number = ds['chr'].iloc[0]\n",
    "    for col in small_list:\n",
    "        mdf = ds[['index',col]]\n",
    "        grouped_df = mdf.groupby(bins)[col].agg(['mean','count'])\n",
    "        grouped_df[f\"num_of_nans_chr{chromo_number}\"] = 25 - grouped_df[\"count\"]\n",
    "        grouped_df[\"indv_name\"] = col\n",
    "        stats_dataframe  = stats_dataframe.append(grouped_df)\n",
    "    return stats_dataframe\n",
    "\n",
    "def remove_high_NaNs(stats_dataframe, chromo_number):\n",
    "    stats_dataframe.loc[stats_dataframe[f\"num_of_nans_chr{chromo_number}\"] > 9, 'mean'] = np.nan\n",
    "    return stats_dataframe\n",
    "\n",
    "def narrow(filt_stats_dataframe): \n",
    "    filt_stats_narrow = filt_stats_dataframe[['indv_name', 'mean']]\n",
    "    filt_stats_narrow['bin'] = filt_stats_narrow.index\n",
    "    filt_stats_wide = filt_stats_narrow.pivot_table(index='bin', columns='indv_name', values='mean', aggfunc='first', dropna=False)\n",
    "    return filt_stats_wide\n",
    "\n",
    "#def cleanup_cols_names(df):\n",
    "    #df.columns = df.columns.str.split('processed/').str[1].tolist()   \n",
    "    #return df\n",
    "\n",
    "def create_bin_template():\n",
    "    col_names = ['level_0', 'first','last']\n",
    "    return pd.DataFrame(columns=col_names)\n",
    "\n",
    "def create_append_template():\n",
    "    col_names = df.columns[2:323]\n",
    "    return pd.DataFrame(columns=col_names)\n",
    "\n",
    "def convert_to_genotypes(template_df):\n",
    "    choiceList = ['BB', 'AA', 'AB']\n",
    "    all_indvs = template_df.columns\n",
    "    for i in all_indvs:\n",
    "        condList = [(template_df[i] > 1.4), (template_df[i] < .6), ((template_df[i] < 1.3) & (template_df[i] > .7))]\n",
    "        template_df[i] = np.select(condList, choiceList)\n",
    "    for i in template_df.columns:\n",
    "        template_df[i].replace('0',np.nan,inplace=True)\n",
    "    return template_df\n",
    "\n",
    "def rename_SNP_bins(ds):\n",
    "    bin_rename = pd.DataFrame(ds['index'].groupby(cut_bins).agg(['first', 'last']).stack())\n",
    "    bin_rename = bin_rename.reset_index()\n",
    "    bin_rename = bin_rename.pivot_table(index='level_0', columns='level_1', values =0,aggfunc='first', dropna=False)\n",
    "    bin_rename = bin_rename.reset_index()\n",
    "    return bin_rename\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data/lg_chromo_9.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-29-baa4727b9fb1>:33: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  filt_stats_narrow['bin'] = filt_stats_narrow.index\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data/lg_chromo_8.csv\n",
      "data/lg_chromo_5.csv\n",
      "data/lg_chromo_14.csv\n",
      "data/lg_chromo_4.csv\n",
      "data/lg_chromo_6.csv\n",
      "data/lg_chromo_7.csv\n",
      "data/lg_chromo_3.csv\n",
      "data/lg_chromo_12.csv\n",
      "data/lg_chromo_13.csv\n",
      "data/lg_chromo_2.csv\n",
      "data/lg_chromo_11.csv\n",
      "data/lg_chromo_10.csv\n",
      "data/lg_chromo_1.csv\n"
     ]
    }
   ],
   "source": [
    "subset_file_list = glob.glob('data/*.csv')\n",
    "ind_list = list_name_indivs(df)\n",
    "\n",
    "template_df = create_append_template()\n",
    "bin_df = create_bin_template()\n",
    "\n",
    "for fil in subset_file_list:\n",
    "    print(fil)\n",
    "    #load subset of dataset that was subset by chromosome\n",
    "    ds = pd.read_csv(fil)\n",
    "    chromo_number = ds['chr'].iloc[0]\n",
    "    #create bins\n",
    "    bins = create_bins(ds)\n",
    "    #create empty stats template\n",
    "    stats_template_df = create_stats_template(ds, chromo_number)\n",
    "    # cut each linkage group into bins of X SNPs each.\n",
    "    cut_bins = cut_by_bins(ds, bins)\n",
    "    # start renaming SNP bins\n",
    "    SNP_bin_rename = rename_SNP_bins(ds)\n",
    "    # create dataframe of stats\n",
    "    stats_dataframe = create_stats_dataframe(ind_list, ds, cut_bins, chromo_number)\n",
    "    # remove dfs with lots of NAs \n",
    "    filt_stats_dataframe = remove_high_NaNs(stats_dataframe, chromo_number)\n",
    "    # pivot magic\n",
    "    filt_stats_wide = narrow(filt_stats_dataframe)\n",
    "    #cleanup col names\n",
    "    #clean_df = cleanup_cols_names(filt_stats_wide)\n",
    "    template_df = template_df.append(filt_stats_wide,ignore_index=True)\n",
    "    bin_df = bin_df.append(SNP_bin_rename, ignore_index=True)\n",
    "#     break\n",
    "#     cleaned_ds.to_csv('final_data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# label the bins by the means of their SNP positions\n",
    "\n",
    "tmpDF1 = pd.DataFrame(columns=['chr','site1', 'chr_', 'site2'])\n",
    "tmpDF1[['chr','site1']] = bin_df['first'].str.split('_', expand=True)\n",
    "tmpDF1[['chr_','site2']] = bin_df['last'].str.split('_', expand=True)\n",
    "\n",
    "tmpDF1['sum'] = pd.to_numeric(tmpDF1['site1']) + pd.to_numeric(tmpDF1['site2'])\n",
    "\n",
    "tmpDF1['mean'] = tmpDF1['sum'] / 2\n",
    "\n",
    "tmpDF1['bin_mean'] = tmpDF1['chr']+\"_\"+tmpDF1['mean'].astype(str)\n",
    "\n",
    "#add the means to the whole spreadsheet\n",
    "template_df.insert(loc=0, column='bin_mean', value=tmpDF1['bin_mean'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add a chr column\n",
    "template_df.insert(loc=0, column='chr', value=tmpDF1['chr'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add a pos column\n",
    "template_df.insert(loc=0, column='pos', value=tmpDF1['mean'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "#i have figured out almost everything except how to replace the values w/AA AB BB, \n",
    "# and I know there is a way to do it without for loops but my brain can't do it right now\n",
    "template_df.to_parquet(\"AHQsd_F2_SNPs_windowed.parquet\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "td = pd.read_parquet(\"AHQsd_F2_SNPs_windowed.parquet\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "td_geno = convert_to_genotypes(td.iloc[:, 3:324])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "td_geno.insert(loc=0, column='bin_mean', value=td['bin_mean'])\n",
    "td_geno.insert(loc=0, column='chr', value=td['chr'])\n",
    "td_geno.insert(loc=0, column='pos', value=td['pos'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "td_geno.to_parquet(\"AHQsd_F2_genotypes_windowed.parquet\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "td_geno.to_csv(\"AHQsd_F2_genotypes_windowed.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>chr</th>\n",
       "      <th>bin_mean</th>\n",
       "      <th>AHQsd_5.01C</th>\n",
       "      <th>AHQsd_5.01D</th>\n",
       "      <th>AHQsd_5.01E</th>\n",
       "      <th>AHQsd_5.01F</th>\n",
       "      <th>AHQsd_5.01G</th>\n",
       "      <th>AHQsd_5.01H</th>\n",
       "      <th>AHQsd_5.02A</th>\n",
       "      <th>...</th>\n",
       "      <th>AHQsd_3.10H</th>\n",
       "      <th>AHQsd_3.11B</th>\n",
       "      <th>AHQsd_3.11C</th>\n",
       "      <th>AHQsd_3.11D</th>\n",
       "      <th>AHQsd_3.11E</th>\n",
       "      <th>AHQsd_3.11G</th>\n",
       "      <th>AHQsd_3.11H</th>\n",
       "      <th>AHQsd_3.12D</th>\n",
       "      <th>AHQsd_3.12E</th>\n",
       "      <th>AHQsd_3.12G</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>94759.0</td>\n",
       "      <td>9</td>\n",
       "      <td>9_94759.0</td>\n",
       "      <td>BB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>124989.0</td>\n",
       "      <td>9</td>\n",
       "      <td>9_124989.0</td>\n",
       "      <td>BB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AB</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>257994.5</td>\n",
       "      <td>9</td>\n",
       "      <td>9_257994.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>675849.5</td>\n",
       "      <td>9</td>\n",
       "      <td>9_675849.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1052261.0</td>\n",
       "      <td>9</td>\n",
       "      <td>9_1052261.0</td>\n",
       "      <td>AB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AB</td>\n",
       "      <td>AB</td>\n",
       "      <td>AB</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AB</td>\n",
       "      <td>AB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1431</th>\n",
       "      <td>11039990.5</td>\n",
       "      <td>1</td>\n",
       "      <td>1_11039990.5</td>\n",
       "      <td>AA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1432</th>\n",
       "      <td>11086062.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1_11086062.0</td>\n",
       "      <td>AB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AB</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AB</td>\n",
       "      <td>AA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1433</th>\n",
       "      <td>11305685.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1_11305685.0</td>\n",
       "      <td>AB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AA</td>\n",
       "      <td>AB</td>\n",
       "      <td>...</td>\n",
       "      <td>AB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1434</th>\n",
       "      <td>11661804.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1_11661804.0</td>\n",
       "      <td>AA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1435</th>\n",
       "      <td>11941314.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1_11941314.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1436 rows × 324 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             pos chr      bin_mean AHQsd_5.01C AHQsd_5.01D AHQsd_5.01E  \\\n",
       "0        94759.0   9     9_94759.0          BB         NaN         NaN   \n",
       "1       124989.0   9    9_124989.0          BB         NaN         NaN   \n",
       "2       257994.5   9    9_257994.5         NaN         NaN         NaN   \n",
       "3       675849.5   9    9_675849.5         NaN         NaN         NaN   \n",
       "4      1052261.0   9   9_1052261.0          AB         NaN         NaN   \n",
       "...          ...  ..           ...         ...         ...         ...   \n",
       "1431  11039990.5   1  1_11039990.5          AA         NaN         NaN   \n",
       "1432  11086062.0   1  1_11086062.0          AB         NaN         NaN   \n",
       "1433  11305685.0   1  1_11305685.0          AB         NaN         NaN   \n",
       "1434  11661804.0   1  1_11661804.0          AA         NaN         NaN   \n",
       "1435  11941314.0   1  1_11941314.0         NaN         NaN         NaN   \n",
       "\n",
       "     AHQsd_5.01F AHQsd_5.01G AHQsd_5.01H AHQsd_5.02A  ... AHQsd_3.10H  \\\n",
       "0            NaN          AA          AA          AA  ...         NaN   \n",
       "1             AB          AA          AA          AA  ...         NaN   \n",
       "2            NaN         NaN          AB         NaN  ...         NaN   \n",
       "3            NaN         NaN          AA         NaN  ...         NaN   \n",
       "4            NaN          AB          AB          AB  ...         NaN   \n",
       "...          ...         ...         ...         ...  ...         ...   \n",
       "1431         NaN         NaN          AA         NaN  ...         NaN   \n",
       "1432         NaN         NaN         NaN          AB  ...         NaN   \n",
       "1433         NaN         NaN          AA          AB  ...          AB   \n",
       "1434         NaN          AA          AA         NaN  ...         NaN   \n",
       "1435         NaN         NaN         NaN         NaN  ...         NaN   \n",
       "\n",
       "     AHQsd_3.11B AHQsd_3.11C AHQsd_3.11D AHQsd_3.11E AHQsd_3.11G AHQsd_3.11H  \\\n",
       "0            NaN          AA         NaN         NaN         NaN         NaN   \n",
       "1            NaN          AA         NaN          AA         NaN         NaN   \n",
       "2            NaN          AB         NaN          AB         NaN         NaN   \n",
       "3            NaN          AA         NaN         NaN         NaN         NaN   \n",
       "4             AB          AB         NaN          AB         NaN         NaN   \n",
       "...          ...         ...         ...         ...         ...         ...   \n",
       "1431         NaN         NaN         NaN          AA         NaN         NaN   \n",
       "1432          AB         NaN         NaN          AB          AA         NaN   \n",
       "1433         NaN          AA         NaN          AB         NaN         NaN   \n",
       "1434         NaN         NaN         NaN         NaN         NaN         NaN   \n",
       "1435         NaN         NaN         NaN         NaN         NaN         NaN   \n",
       "\n",
       "     AHQsd_3.12D AHQsd_3.12E AHQsd_3.12G  \n",
       "0            NaN         NaN         NaN  \n",
       "1            NaN         NaN         NaN  \n",
       "2            NaN         NaN         NaN  \n",
       "3            NaN         NaN         NaN  \n",
       "4            NaN         NaN         NaN  \n",
       "...          ...         ...         ...  \n",
       "1431         NaN         NaN         NaN  \n",
       "1432         NaN         NaN         NaN  \n",
       "1433         NaN         NaN         NaN  \n",
       "1434         NaN         NaN         NaN  \n",
       "1435         NaN         NaN         NaN  \n",
       "\n",
       "[1436 rows x 324 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "td_geno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use apply to make a table with the lists of genotype counts per individual\n",
    "td_geno_counts_inds = td_geno.iloc[:,3:324].apply(pd.Series.value_counts, axis=0).T\n",
    "\n",
    "# use apply to make a table with the lists of genotype counts per site\n",
    "td_geno_counts_sites = td_geno.iloc[:,3:324].apply(pd.Series.value_counts, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "td_geno_counts_sites.insert(loc=0, column='bin_mean', value=td_geno['bin_mean'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "site_ratios = pd.DataFrame(columns=['AA','AB','BB'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "site_ratios['AA'] = np.divide(td_geno_counts_sites['AA'],\n",
    "                              td_geno_counts_sites['total'])\n",
    "site_ratios['AB'] = np.divide(td_geno_counts_sites['AB'],\n",
    "                              td_geno_counts_sites['total'])\n",
    "site_ratios['BB'] = np.divide(td_geno_counts_sites['BB'],\n",
    "                              td_geno_counts_sites['total'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "cannot insert pos, already exists",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-89-07ff03e27687>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m#site_ratios.insert(loc=0, column='bin_mean', value=td_geno['bin_mean'])\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0msite_ratios\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minsert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumn\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'pos'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtd_geno\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'pos'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0msite_ratios\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minsert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumn\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'chr'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtd_geno\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'chr'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/miniconda3/lib/python3.8/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36minsert\u001b[0;34m(self, loc, column, value, allow_duplicates)\u001b[0m\n\u001b[1;32m   3619\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_ensure_valid_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3620\u001b[0m         \u001b[0mvalue\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sanitize_column\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcolumn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbroadcast\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3621\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_mgr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minsert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mallow_duplicates\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mallow_duplicates\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3622\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3623\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0massign\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;34m\"DataFrame\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/miniconda3/lib/python3.8/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36minsert\u001b[0;34m(self, loc, item, value, allow_duplicates)\u001b[0m\n\u001b[1;32m   1175\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mallow_duplicates\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mitem\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1176\u001b[0m             \u001b[0;31m# Should this be a different kind of error??\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1177\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"cannot insert {item}, already exists\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1178\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1179\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: cannot insert pos, already exists"
     ]
    }
   ],
   "source": [
    "#site_ratios.insert(loc=0, column='bin_mean', value=td_geno['bin_mean'])\n",
    "site_ratios.insert(loc=0, column='pos', value=td_geno['pos'])\n",
    "site_ratios.insert(loc=0, column='chr', value=td_geno['chr'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       172.0\n",
       "1       435.0\n",
       "2       312.0\n",
       "3        86.0\n",
       "4       262.0\n",
       "        ...  \n",
       "1431    122.0\n",
       "1432    164.0\n",
       "1433    134.0\n",
       "1434    186.0\n",
       "1435      0.0\n",
       "Length: 1436, dtype: float64"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "td_geno_counts_sites.sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>bin_mean</th>\n",
       "      <th>AA</th>\n",
       "      <th>AB</th>\n",
       "      <th>BB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9</td>\n",
       "      <td>94759.0</td>\n",
       "      <td>9_94759.0</td>\n",
       "      <td>0.651163</td>\n",
       "      <td>0.255814</td>\n",
       "      <td>0.093023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9</td>\n",
       "      <td>124989.0</td>\n",
       "      <td>9_124989.0</td>\n",
       "      <td>0.310345</td>\n",
       "      <td>0.131034</td>\n",
       "      <td>0.058621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9</td>\n",
       "      <td>257994.5</td>\n",
       "      <td>9_257994.5</td>\n",
       "      <td>0.004808</td>\n",
       "      <td>0.495192</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9</td>\n",
       "      <td>675849.5</td>\n",
       "      <td>9_675849.5</td>\n",
       "      <td>0.488372</td>\n",
       "      <td>0.511628</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9</td>\n",
       "      <td>1052261.0</td>\n",
       "      <td>9_1052261.0</td>\n",
       "      <td>0.015267</td>\n",
       "      <td>0.984733</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1431</th>\n",
       "      <td>1</td>\n",
       "      <td>11039990.5</td>\n",
       "      <td>1_11039990.5</td>\n",
       "      <td>0.311475</td>\n",
       "      <td>0.688525</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1432</th>\n",
       "      <td>1</td>\n",
       "      <td>11086062.0</td>\n",
       "      <td>1_11086062.0</td>\n",
       "      <td>0.353659</td>\n",
       "      <td>0.621951</td>\n",
       "      <td>0.024390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1433</th>\n",
       "      <td>1</td>\n",
       "      <td>11305685.0</td>\n",
       "      <td>1_11305685.0</td>\n",
       "      <td>0.119403</td>\n",
       "      <td>0.850746</td>\n",
       "      <td>0.029851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1434</th>\n",
       "      <td>1</td>\n",
       "      <td>11661804.0</td>\n",
       "      <td>1_11661804.0</td>\n",
       "      <td>0.924731</td>\n",
       "      <td>0.075269</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1435</th>\n",
       "      <td>1</td>\n",
       "      <td>11941314.0</td>\n",
       "      <td>1_11941314.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1436 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr         pos      bin_mean        AA        AB        BB\n",
       "0      9     94759.0     9_94759.0  0.651163  0.255814  0.093023\n",
       "1      9    124989.0    9_124989.0  0.310345  0.131034  0.058621\n",
       "2      9    257994.5    9_257994.5  0.004808  0.495192       NaN\n",
       "3      9    675849.5    9_675849.5  0.488372  0.511628       NaN\n",
       "4      9   1052261.0   9_1052261.0  0.015267  0.984733       NaN\n",
       "...   ..         ...           ...       ...       ...       ...\n",
       "1431   1  11039990.5  1_11039990.5  0.311475  0.688525       NaN\n",
       "1432   1  11086062.0  1_11086062.0  0.353659  0.621951  0.024390\n",
       "1433   1  11305685.0  1_11305685.0  0.119403  0.850746  0.029851\n",
       "1434   1  11661804.0  1_11661804.0  0.924731  0.075269       NaN\n",
       "1435   1  11941314.0  1_11941314.0       NaN       NaN       NaN\n",
       "\n",
       "[1436 rows x 6 columns]"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "site_ratios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>bin_mean</th>\n",
       "      <th>AA</th>\n",
       "      <th>AB</th>\n",
       "      <th>BB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9</td>\n",
       "      <td>94759.0</td>\n",
       "      <td>9_94759.0</td>\n",
       "      <td>0.325581</td>\n",
       "      <td>0.127907</td>\n",
       "      <td>0.046512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9</td>\n",
       "      <td>124989.0</td>\n",
       "      <td>9_124989.0</td>\n",
       "      <td>0.206897</td>\n",
       "      <td>0.087356</td>\n",
       "      <td>0.039080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9</td>\n",
       "      <td>257994.5</td>\n",
       "      <td>9_257994.5</td>\n",
       "      <td>0.003205</td>\n",
       "      <td>0.330128</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9</td>\n",
       "      <td>675849.5</td>\n",
       "      <td>9_675849.5</td>\n",
       "      <td>0.244186</td>\n",
       "      <td>0.255814</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9</td>\n",
       "      <td>1052261.0</td>\n",
       "      <td>9_1052261.0</td>\n",
       "      <td>0.007634</td>\n",
       "      <td>0.492366</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1431</th>\n",
       "      <td>1</td>\n",
       "      <td>11039990.5</td>\n",
       "      <td>1_11039990.5</td>\n",
       "      <td>0.155738</td>\n",
       "      <td>0.344262</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1432</th>\n",
       "      <td>1</td>\n",
       "      <td>11086062.0</td>\n",
       "      <td>1_11086062.0</td>\n",
       "      <td>0.176829</td>\n",
       "      <td>0.310976</td>\n",
       "      <td>0.012195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1433</th>\n",
       "      <td>1</td>\n",
       "      <td>11305685.0</td>\n",
       "      <td>1_11305685.0</td>\n",
       "      <td>0.059701</td>\n",
       "      <td>0.425373</td>\n",
       "      <td>0.014925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1434</th>\n",
       "      <td>1</td>\n",
       "      <td>11661804.0</td>\n",
       "      <td>1_11661804.0</td>\n",
       "      <td>0.462366</td>\n",
       "      <td>0.037634</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1435</th>\n",
       "      <td>1</td>\n",
       "      <td>11941314.0</td>\n",
       "      <td>1_11941314.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1436 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr         pos      bin_mean        AA        AB        BB\n",
       "0      9     94759.0     9_94759.0  0.325581  0.127907  0.046512\n",
       "1      9    124989.0    9_124989.0  0.206897  0.087356  0.039080\n",
       "2      9    257994.5    9_257994.5  0.003205  0.330128       NaN\n",
       "3      9    675849.5    9_675849.5  0.244186  0.255814       NaN\n",
       "4      9   1052261.0   9_1052261.0  0.007634  0.492366       NaN\n",
       "...   ..         ...           ...       ...       ...       ...\n",
       "1431   1  11039990.5  1_11039990.5  0.155738  0.344262       NaN\n",
       "1432   1  11086062.0  1_11086062.0  0.176829  0.310976  0.012195\n",
       "1433   1  11305685.0  1_11305685.0  0.059701  0.425373  0.014925\n",
       "1434   1  11661804.0  1_11661804.0  0.462366  0.037634       NaN\n",
       "1435   1  11941314.0  1_11941314.0       NaN       NaN       NaN\n",
       "\n",
       "[1436 rows x 6 columns]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "site_ratios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
